libraries/lib-cloud-audiocom/OAuthService.cpp

```diff
@@ -0,0 +1,370 @@
+const std::string_view passwordPrefix = "password=";
+const std::string_view tokenPrefix = "token=";
+const std::string_view authorizationCodePrefix = "authorization_code=";
+
+void WriteCommonFields(
+   rapidjson::Document& document, std::string_view grantType, std::string_view scope)
+{
+   using namespace rapidjson;
+
+   document.AddMember(
+      "grant_type", StringRef(grantType.data(), grantType.size()),
+      document.GetAllocator());
+
+   const auto clientID = GetServiceConfig().GetOAuthClientID();
+
+   document.AddMember(
+      "client_id", StringRef(clientID.data(), clientID.size()),
+      document.GetAllocator());
+
+   document.AddMember(
+      "client_secret", StringRef("shKqnY2sLTfRK7hztwzNEVxnmhJfOy1i"),
+      document.GetAllocator());
+
+   document.AddMember(
+      "scope", StringRef(scope.data(), scope.size()), document.GetAllocator());
+}
+
+bool IsPrefixed(std::string_view hay, std::string_view prefix)
+{
+   if (hay.length() < prefix.length())
+      return false;
+
+   return std::equal(
+      hay.begin(), hay.begin() + prefix.length(), prefix.begin(), prefix.end(),
+      [](auto a, auto b) { return std::tolower(a) == b; });
+}
+
+} // namespace
+
+void OAuthService::ValidateAuth(
+   std::function<void(std::string_view)> completedHandler)
+{
+   if (HasAccessToken() || !HasRefreshToken())
+   {
+      if (completedHandler)
+         completedHandler(GetAccessToken());
+      return;
+   }
+
+   AuthoriseRefreshToken(GetServiceConfig(), std::move(completedHandler));
+}
+
+void OAuthService::HandleLinkURI(
+   std::string_view uri, std::function<void(std::string_view)> completedHandler)
+{
+   if (!IsPrefixed(uri, uriPrefix))
+   {
+      if (completedHandler)
+         completedHandler({});
+      return;
+   }
+
+   // It was observed, that sometimes link is passed as audacity://link/
+   // This is valid from URI point of view, but we need to handle it separately
+   const auto argsStart = uri.find("?");
+
+   if (argsStart == std::string_view::npos)
+   {
+      if (completedHandler)
+         completedHandler({});
+      return;
+   }
+
+   // Length is handled in IsPrefixed
+   auto args = uri.substr(argsStart + 1);
+
+   std::string_view token;
+   std::string_view username;
+   std::string_view password;
+   std::string_view authorizationCode;
+
+   while (!args.empty())
+   {
+      const auto nextArg = args.find('&');
+
+      const auto arg = args.substr(0, nextArg);
+      args = nextArg == std::string_view::npos ? "" : args.substr(nextArg + 1);
+
+      if (IsPrefixed(arg, usernamePrefix))
+         username = arg.substr(usernamePrefix.length());
+      else if (IsPrefixed(arg, passwordPrefix))
+         password = arg.substr(passwordPrefix.length());
+      else if (IsPrefixed(arg, tokenPrefix))
+         token = arg.substr(tokenPrefix.length());
+      else if (IsPrefixed(arg, authorizationCodePrefix))
+         authorizationCode = arg.substr(authorizationCodePrefix.length());
```

Comment:
Do you want to defend against a string the defines one of the arguments more than once?

Reply:
Nope